backend/src/main/scala/sbt/internal/inc/bloop/internal/ThreadSafeAnalysisCallback.scala

```diff
@@ -0,0 +1,398 @@
+package sbt.internal.inc.bloop.internal
+
+import java.io.File
+import java.{util => ju}
+
+import xsbti.api.AnalyzedClass
+import xsbti.compile.analysis.ReadStamps
+import xsbti.compile.Output
+import xsbti.compile.IncOptions
+import xsbti.api.SafeLazyProxy
+import xsbt.api.HashAPI
+import xsbti.api.ClassLike
+import xsbti.api.NameHash
+import xsbti.Problem
+import xsbti.api.InternalDependency
+import xsbti.api.ExternalDependency
+import xsbti.api.DependencyContext
+import xsbti.Position
+import xsbti.Severity
+import xsbti.api.DefinitionType
+import xsbti.UseScope
+import xsbti.api.Companions
+import xsbt.api.APIUtil
+import xsbt.api.NameHashing
+import xsbti.compile.ClassFileManager
+
+import sbt.internal.inc.Incremental
+import sbt.util.InterfaceUtil
+import sbt.internal.inc.UsedName
+import sbt.internal.inc.Analysis
+import sbt.internal.inc.Compilation
+import sbt.internal.inc.SourceInfos
+
+import bloop.CompileMode
+import xsbti.compile.Signature
+
+/**
+ * This class provides a thread-safe implementation of `xsbti.AnalysisCallback` which is required to compile with the
+ * Triplequote Hydra compiler.
+ *
+ * In essence, the implementation is a merge of the `BloopAnalysisCallback` with the original sbt default `AnalysisCallback`
+ * implementation (https://github.com/sbt/zinc/blob/develop/internal/zinc-core/src/main/scala/sbt/internal/inc/Compile.scala),
+ * which is already thread-safe.
+ *
+ * IMPORTANT: All modifications made to BloopAnalysisCallback` must be replicated here.
+ */
+final class ThreadSafeAnalysisCallback(
```

Comment:
A better name would be `ConcurrentAnalysisCallback`. I'm happy to make the change if you'd like so.

Reply:
Alright, I'll do it if there are other changes that are needed (which I'm not sure there will be). So, this shouldn't be blocking a merge if you are happy with the PR.